### ISSUE TITLE
[consistent-return] false positive when returning a void/undefined value

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#1277]
```TS
/* eslint 'consistent-return': 'error' */

declare function returnsVoid(): void;
function foo(condition: boolean): void {
//             ^^^ Expected a return value at the end of of function 'foo'. (consistent-return
  if (condition) {
    return returnsVoid();
  }

  console.log('no return');
}
```

**Additional Info**

https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50ICdjb25zaXN0ZW50LXJldHVybic6ICdlcnJvcicgKi9cblxuZnVuY3Rpb24gZm9vKGNvbmRpdGlvbikge1xuICBpZiAoY29uZGl0aW9uKSB7XG4gICAgcmV0dXJuIHJldHVybnNWb2lkKCk7XG4gIH1cblxuICBjb25zb2xlLmxvZygnbm8gcmV0dXJuJyk7XG59XG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjksInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19

----

We should be able to make `consistent-return` type-aware, so that it doesn't false positive on void/undefined returning functions.

**Comment by CatsMiaow** [comment]
![image](https://user-images.githubusercontent.com/1300172/74209353-b874f980-4cca-11ea-81b8-f22d4387084b.png)

**Comment by piranna** [comment]
Any update on this?

**Comment by bradzacher** [comment]
With any issue in opened in this project - it either has a visible progress in the form of an attached PR, or it has no progress.

We are a community run project. The **volunteer** maintainers spend most of their time triaging issues and reviewing PRs. This means that most issues *will not progress* unless a member of the community steps up and **champions** it.

If this issue is important to you - consider being that champion.

If not - please just subscribe to the issue and wait patiently.
Commenting asking for status updates does not bump issue priority in any way and just serves to spam ***everyone*** subscribed to the issue.

**Comment by Josh-Cena** [comment]
Note that @CatsMiaow's reply illustrates a different case which I don't think we should seek to fix. We only want to treat `return expression;` where `expression` is `void`/`undefined` the same as `return;`, but leave other cases where `expression` is not `void` as-is, because they are not "consistent" and defeats the purpose of distinguishing between `void`-returning and non-`void`-returning. This pattern is also unsafe:

```ts
function f(): string | void {
  if (Math.random() > 0.5) return "a";
  return;
}

function another(y: string | undefined) {}

const x = f();
another(x);
// Argument of type 'string | void' is not assignable to parameter of type 'string | undefined'.
//  Type 'void' is not assignable to type 'string | undefined'
```

This is because `void` as a type effectively behaves like `unknown`, so you could actually not do anything useful with it without manually narrowing the type anyway. In this case you absolutely should declare `f` as returning `string | undefined` instead.

**Comment by piranna** [comment]
> We only want to treat `return expression;` where `expression` is `void`/`undefined` the same as `return;`, but leave other cases where `expression` is not `void` as-is

Totally agree.


> `void` as a type effectively behaves like `unknown`

I though `void` as a type worked as `undefined`, since `void` statement always return `undefined`, the same way that's being used `void 0` to generate `undefined`. Can you point me to some docs about that equivalence to `unknown`?

**Comment by Josh-Cena** [comment]
https://github.com/microsoft/TypeScript/issues/42709 is a fun read.

Consider this simple example:

```ts
const a: () => void = () => 1;
const b = a(); // b is void
```

TS lets you do this because `void` simply means "I will never use the return type so I don't care what it is". The return type `void` is treated as `undefined` internally (within the function), but as `unknown` externally (when doing assignability checks).

**Comment by bradzacher** [comment]
The `void` type is a special type that doesn't behave in like `undefined` always. It has some special rules around it that dictate what is assignable to it.

For example - when it is the return type of the function - that function body can either have no return, a return with no expression, or a return with an expression whose type is `void` or `undefined`. 

When used as the direct return type for a function type and that function type is used in a receiving postion (eg things are being assigned to that type) then it acts like `unknown` and thus any function return type may be assigned to `void`.

When used as an argument type - `void` and `undefined` are not assignable - nor is any other type. It acts mor like a nominal type here.

When used in a union type in a function return type for a function type that is used in a receiving postion then it acts like `undefined`, and not like `unknown` as before.

All in all its a really complex type and ultimately you should avoid using it in any location that is not the direct return type location for a function or function type - we have lint rule for that [`no-invalid-void-type`](https://typescript-eslint.io/rules/no-invalid-void-type/)

**Comment by piranna** [comment]
Thanks both for clarification. So, one thing is `void` used as a JavaScript statement, that always return `undefined` to whatever is given, and another thing is `void` as Typescript type, that depending of where it's being used, represents `undefined` or `unknown`, is that correct?

**Comment by bradzacher** [comment]
Correct - the `void` operator (eg `void foo`) always returns `undefined`.

And the `void` type which sometimes is a nominal type, sometimes means `undefined` and sometimes means `unknown`.

**Comment by piranna** [comment]
Thank you @bradzacher :-)
